memcache/memcache.go

```diff
@@ -711,3 +729,244 @@ func (c *Client) incrDecr(verb, key string, delta uint64) (uint64, error) {
+	Invalidate bool
+
+	// use if you want to get the key as a part of the response
+	// equivalent to the k flag
+	ReturnKeyInResponse bool
+
+	// equivalent to the q flag
+	UseNoReplySemanticsForResponse bool
+
+	// use if only want to store a value if the supplied token matches the current CAS token of the item
+	// equivalent to the C<token> flag
+	CompareCasTokenToUpdateValue *uint64
+
+	// use if you want to set client flags to a token
+	// equivalent to the F<token> flag
+	ClientFlagToken *uint32
+
+	// use if you want to consume a token and copy back with a response
+	// equivalent to the O<token> flag
+	OpaqueToken *string
+
+	// use if you want to set the TTL for the item
+	// equivalent to the T<token> flag
+	UpdateTTLToken *uint32
+
+	// use if you want to switch modes
+	// E: "add" command. LRU bump and return NS if item exists, else add
+	// A: "append" command. If item exists, append the new value to its data
+	// P: "prepend" command. If item exists, prepend the new value to its data
+	// R: "replace" command. Set only if item exists, replace its value
+	// S: "set" command. The default mode, added for completeness
+	// equivalent to the M<token> flag
+	SetModeToken *string
+}
+
+func (c *Client) MetaSet(metaItem *MetaSetItem) (metaDataResponse *MetaResponseMetadata, err error) {
+	err = c.onMetaItem(metaItem, (*Client).metaSet, func(metaData *MetaResponseMetadata) { metaDataResponse = metaData })
+	return
+}
+
+func (c *Client) onMetaItem(item *MetaSetItem, fn func(*Client, *bufio.ReadWriter, *MetaSetItem, func(metaData *MetaResponseMetadata)) error, cb func(metaData *MetaResponseMetadata)) error {
+	addr, err := c.selector.PickServer(item.Key)
+	if err != nil {
+		return err
+	}
+	cn, err := c.getConn(addr)
+	if err != nil {
+		return err
+	}
+	defer cn.condRelease(&err)
+	if err = fn(c, cn.rw, item, cb); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *Client) metaSet(rw *bufio.ReadWriter, item *MetaSetItem, cb func(metaData *MetaResponseMetadata)) error {
+	return c.processMetaSet(rw, item, cb)
+}
+
+func (c *Client) processMetaSet(rw *bufio.ReadWriter, item *MetaSetItem, cb func(*MetaResponseMetadata)) error {
+	if !legalKey(item.Key) {
+		return ErrMalformedKey
+	}
+
+	command := c.parseFlagsForMetaSet(item)
+
+	var err error
+	_, err = fmt.Fprintf(rw, command)
+
+	if err != nil {
+		return nil
+	}
+	if _, err = rw.Write(item.Value); err != nil {
+		return err
+	}
+	if _, err := rw.Write(crlf); err != nil {
+		return err
+	}
+	if err := rw.Flush(); err != nil {
+		return err
+	}
+	response, err := rw.ReadSlice('\n')
+	if err != nil {
+		return err
+	}
+	switch {
+	case bytes.HasPrefix(response, metaResultStored):
+		// the first two characters are being processed in this switch-case block, the other cases are all errors
+		// so we don't need to save them into the meta data object
+		responseMetadataComponents := strings.Fields(string(response))[1:]
+		var metaResponse *MetaResponseMetadata
```

Comment:
Since we are using MetaResponseMetadata for all meta commands can we create the struct even if the user does not select any response flag?. The reason for this is in meta get I will always return the item value and i have to create Meta response metadata.

Reply:
I think for the sake of consistency, we should always return the struct. Even if none of the values are filled- as a client, methods can get confusing if sometimes I get an object back, sometimes I don't. What are your thoughts on that?

Reply:
Yes. Lets always return the struct. It also makes my Meta get implementation easy.